### ISSUE TITLE
HDDL - NPE when using disjunctive and existential preconditions (imply .... (exists .... ))

### ISSUE THREAD

**Issue by o-fir** [pellierd/pddl4j#93]
Imagine we want to express the following constraint in a dummy scheduling problem:
_For a task **t** to be planned at timeslot **s_i**, all tasks that are predecessors of task **t** should be planned in some timeslot **t_j**, where **t_j < t_i**_

When using non-temporal PDDL, we require the following predicates:

- `(task_only_after ?t2 - task ?t1 - task)` - indicating that task **t2** should happen after task **t1**
- `(is_before ?s1 - timeslot ?s2 - timeslot)` - indicating that timeslot **s1** "happens" before timeslot **s2**

With these predicates, for a task **t** to be planned at timeslot **s**, the following constraint should be satisfied:
`(forall (?tPrime - task) (imply (task_only_after ?t ?tPrime) (exists (?sPrime - timeslot) (and (task_planned ?tPrime ?sPrime) (is_before ?sPrime ?s)))))`
_or in other words:
if there is some task **tPrime**, which should precede task **t**, then there should exist a timeslot **sPrime**, during which **tPrime** is planned, and which happens before timeslot **s**._

This, however, results in an error when calling _problem.instatiate()_:
```
Exception in thread "main" java.lang.IndexOutOfBoundsException: Index: 42, Size: 27
        at java.util.ArrayList.rangeCheck(ArrayList.java:659)
        at java.util.ArrayList.get(ArrayList.java:435)
        at fr.uga.pddl4j.problem.FinalizedProblem.finalizeMethods(FinalizedProblem.java:1585)
        at fr.uga.pddl4j.problem.HTNProblem.finalization(HTNProblem.java:206)
        at fr.uga.pddl4j.problem.AbstractProblem.instantiate(AbstractProblem.java:438)
```

Rephrasing **exists** as a **(forall ... (imply ... ))** solves the problem, although the two constraints aren't equivalent
`(forall (?tPrime - task) (imply (task_only_after ?t ?tPrime) (forall (?sPrime - timeslot)(imply (task_planned ?tPrime ?sPrime) (is_before ?sPrime ?s)))))`

**DOMAIN:**
```

;NOTE - CURRENT PLANNER DOESNT SUPPORT NEGATIVE GOALS (i.e. (not PREDICATE) in goals)

(define (domain bassetti-javascript)

(:requirements 
    :hierarchy
	  :typing
	  :method-preconditions
    :disjunctive-preconditions ; imply
    :negative-preconditions
    :universal-preconditions ; forall
    :conditional-effects ;forall
    :existential-preconditions ; exists
)

(:types     
    timeslot
    resource
    task
)

(:predicates 
 (task_planned ?t - task ?s - timeslot)
 (compatible ?t - task ?r - resource)
 (resource_available ?r - resource ?s - timeslot)
 (deadline ?t - task ?s - timeslot) ; Constant. represents deadlines. If timeslot is after deadline, the predicate is false, otherwise - false. 
 (task_only_after ?t2 - task ?t1 - task) ;t2 can be executed only after t1
 (is_before ?s1 - timeslot ?s2 - timeslot)
 (has_dependencies ?t - task)
)

(:task do_task
  :parameters (?t - task))

; CASE - TASK SHOULD BE PLANNED
; subcase - all requirements are fulfilled. In which case - perform the task
(:method m0_do_task_w_dependencies
  :parameters (?t - task ?r - resource ?s - timeslot)
  :task (do_task ?t)
  :precondition (and  (compatible ?t ?r) (not(deadline ?t ?s)) (has_dependencies ?t) 
  (forall (?tPrime - task) (imply (task_only_after ?t ?tPrime) (exists (?sPrime - timeslot) (and (task_planned ?tPrime ?sPrime) (is_before ?sPrime ?s)))))
  ) ; task not done; task-resource compatible; all deadlines are satisfied)
  :ordered-subtasks(and (t1 (schedule_task ?t ?r ?s)) )
)

(:method m1_do_task_no_dependencies
  :parameters (?t - task ?r - resource ?s - timeslot)
  :task (do_task ?t)
  :precondition (and  (compatible ?t ?r) (not(deadline ?t ?s)) (not(has_dependencies ?t) )
  ) ; task not done; task-resource compatible; all deadlines are satisfied)
  :ordered-subtasks(and (t1 (schedule_task ?t ?r ?s)) )
)

; CASE - TASK ALREADY PLANNED
(:method m2_do_task_nop
  :parameters (?t - task ?s - timeslot)
  :task (do_task ?t)
  :precondition (and (task_planned ?t ?s ))
  :ordered-subtasks(and (t1 (nop)) )
)

; schedule task that's not another task's parent
(:action schedule_task
  :parameters (?t - task ?r - resource ?s - timeslot)
  :precondition (and (not(task_planned ?t ?s )) 
  (resource_available ?r ?s))
  :effect(and (task_planned ?t ?s) (not(resource_available ?r ?s))
  ;(task_performed ?t ?s))
))

(:action nop
  :parameters ()
  :precondition ()
  :effect ()
)
)
```

**PROBLEM:**
```
(define (problem test)
(:domain bassetti-javascript)

;RESOURCES: 18

;TASKS: 56

;TIMESLOTS: 5

;---------------- Facts -----------------------
(:objects
 t0 t1 t2 - timeslot
  res_1 res_2 - resource
  task_1 task_2 task_3 task_4 - task
 )

;--------------- Initial State -----------------
 (:htn
     :parameters()
     :subtasks (and
     (task0(do_task task_4))
     (task1(do_task task_2))
     (task2(do_task task_3))
     (task3(do_task task_1))

      )
   )
(:init
  ; declaring time management
  (is_before t0 t1)
  (is_before t0 t2)
  (is_before t1 t2)
  ; declaring deadlines
  ;(deadline task_1 t2 )
  ;(deadline task_2 t2 )
  ;(deadline task_3 t2 )
  ;(deadline task_4 t2 )
  ; task dependencies
  
  (task_only_after task_3 task_4)
  (has_dependencies task_3)
  ;(task_only_after task_1 task_2)
  ;(has_dependencies task_1)

 ;task-resource compatibility
  ;res 1
  (compatible task_1 res_1)
  (compatible task_2 res_1)
  (compatible task_3 res_1)
  (compatible task_4 res_1)
  ;res 2
  (compatible task_1 res_2)
  (compatible task_2 res_2)
  (compatible task_3 res_2)
  (compatible task_4 res_2)
  

 ;resource properties
  (resource_available res_1 t0)
  (resource_available res_1 t1)
  (resource_available res_1 t2)
  (resource_available res_2 t0)
  (resource_available res_2 t1)
  (resource_available res_2 t2)
)



)
```

**Comment by pellierd** [comment]
There is a bug when methods have disjunctive preconditions. I will investigate further.

**Comment by pellierd** [comment]
The problem is solved. Disjunctive preconditions in methods produced invalid indexes in the task resolvers table of the problem.

The method getTaskResolvers() returns the relevant operators for all the tasks of the problem and null if the problem is not hierarchical.  Warning a task may have many resolvers event primitives tasks.
 
The resolvers returned are indexes of operators. To get the list of resolvers of a specific task _t_ just write:      `List<Integer> resolvers = problem.getTaskResolvers().get(t)`
 
Two case must be considered:

1. If the task {@code t} is primitive, i.e., `problem.getTask(t).isPrimtive()` returns true, the list of resolvers contains either indexes of actions either indexes of durative actions. If the index is positive the index represents an action. To get the corresponding action just use `code problem.getActions(index)`. If the index is negative the index represents a durative action. To get the corresponding durative action just use `problem.getDurativeActions(-index - 1)`.
2. Symmetrically, if the task `t` is compound, i.e., `problem.getTask(t).isCompound()`returns true, the list of resolvers contains either indexes of method either indexes of durative methods. If the index is positive the index represents a method. To get the corresponding method just use `problem.getMethods(index)`. If the index is negative the index represents a durative method. To get the corresponding durative method just use `problem.getDurativeMethods(-index - 1)`.
